### ISSUE TITLE
fs/local: implement hashsum cache

### ISSUE THREAD

**Issue by jediry** [rclone/rclone#949]
I think a nice feature would be the ability to generate a cache of checksums for local files, which could be used to accelerate subsequent calls to rclone. This would make the following sequence of operations much faster, by avoiding redundant checksum computations on the local side:

> **rclone check --checksum cloud:/path /local/path**
> [user thinks about why the files might be different, and decides what to do]
> **rclone copy some files**
> [repeat as necessary]
> **rclone check --checksum cloud:/path /local/path**
> [see where we stand now]

I'm thinking that this might be accomplished with a single command-line argument..something like --local-file-checksum-cache=path/to/database/file. If the file is missing, it's created, and in either case is written out when rclone terminates. The checksum DB would cache the file's size and mod-time along with the checksum, and when the checksum is needed for a local file, the cached one is used if the size and mod-time match.

Note that the cache file would potentially need to cache both MD5 and SHA1 checksums, and there should also be a way to disable reading the cache (e.g., if I believe my filesystem may have been damaged, and want to re-download pristine versions of any corrupted files, then I don't want to trust the local checksum cache).

See also related issues #725, #626 and #157.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by monroe-74** [comment]
This is a great idea, thank you for bringing this up.

<blockquote><p>there should also be a way to disable reading the cache</p></blockquote>

I would think that it would be sufficient for the user to refrain from using the command you suggested ("local-file-checksum-cache"). If the user refrains from using that command, then rclone will behave as it does currently. That is, it will recalculate all checksums, which is the same thing as "disable reading the cache," right?

<blockquote><p>then I don't want to trust the local checksum cache</p></blockquote>

In any situation where the user feels this way, then aside from not using the command, they are also free to trash the checksum file. Then use the command again, and the local checksum cache will be recreated.

Your basic design for how this works makes perfect sense, and it's probably all that's needed. Unless there's something I misunderstand about this part of your comment.

**Comment by jediry** [comment]
Yes, you're correct...so long as the cache file is opt-in, no extra option is needed to disable it. I only mention this scenario in case the person who implements this chooses to instead make this an on-by-default, opt-out thing. The point is, there are cases where one doesn't want the cache.

**Comment by monroe-74** [comment]
<blockquote><p>I only mention this scenario in case the person who implements this chooses to instead make this an on-by-default, opt-out thing.</p></blockquote>

OK, thanks for explaining, I see what you mean. But even in that case, any user who is worried about a corrupted cache file is always free to trash it, and that would always force rclone to recreate it, right? Then again, if for some reason a user worries that the local file is always subject to corruption for some reason, then remembering to delete it every time would be a major nuisance.

Anyway, I like your original suggestion of making this feature happen only when explicitly invoked. In my opinion, it should not be on-by-default. People upgrading to a new rclone version should not suddenly see a new cache file they might not expect or need.

**Comment by jediry** [comment]
I agree that my original opt-in design seems best, as it avoids surprising users.

**Comment by ncw** [comment]
I'd propose implementing it like this.

Use a simple Key/Value database - [bolt](https://github.com/boltdb/bolt) looks like the leading contender here.  Using an actual DB rather than a text file will mean we don't have to load the entire database into memory and we can update it as we go along.

Use the absolute value of the paths as a key - this will mean that you can use the cache for any local file accesses, not just ones from a specific directory.

As the values we need to store
  * modification time
  * sha1sum
  * md5sum

I'd serialize these into JSON probably.

We'd use this DB
  * when reading a Hash we'd look in the DB first
  * when we've calculated a hash we'd stuff it in the DB

Potential issues
  * garbage collection - if you move a lot of files about there will be files in the DB which don't have a corresponding file - I haven't thought of a mechanism for removing these.  It might not be a big problem though.
  * concurrent use. - bolt (and most key value databases) only allow one writer at a time to the db so you'd only be able to use one rclone processes at once.

**Comment by monroe-74** [comment]
Hi Nick, thank you for the work you are putting into this.

<blockquote><p>garbage collection</p></blockquote>

A suggestion: at least for now, just let the user worry about this. As a user, I'm well-aware of the existence of this file, because you won't create it unless I explicitly ask you to do so. It's easy enough for the user to pay attention to where it is, and how big/old/stale it's getting. Periodically, I should trash that file and let a new one be built, but the optimum schedule for that is very dependent on the specifics of my situation. For example, I might know my data well enough to be sure that it's rare for files to be moved or deleted.

Later on, if eventually there are many people using this feature, then you will be in a position to notice the presence (or absence) of many people reporting problems that are a result of garbage accumulating in the file. Then enhancements can be made, and the nature of the problems that are reported (if any) will be a helpful framework for designing those enhancements. Like you said, it might not be a big problem, ever, so no point worrying about it right now.

For now, how about you just put a statement like this in the documentation for this new command: "If you have many files that are moved or deleted, this checksum file will eventually contain obsolete information, so it could be a good idea to periodically throw this file away. Then you can let rclone build a new one which will contain only current information."

<blockquote><p>concurrent use … you'd only be able to use one rclone processes at once</blockquote><p>

rclone currently allows me to create multiple log files with unique names. Then I can run multiple jobs concurrently, and each log file will reflect only the specific job which owns that file. Could checksum files work the same way? Then as a user it's up to me to decide how many of these files should exist, and what they should be called, and where they should live.

The documentation could say something like this: "You can run multiple rclone instances concurrently, but if you are using checksum files, then each instance must have its own."

When a job starts, if I ask you to use a checksum file, you check and see if the file I specified is currently in use. If it's in use, you stop the job with an error. A helpful message would be something like this: "The checksum file you specified is currently in use."

<blockquote><p>is bolt available on all the platforms rclone builds for</blockquote><p>

My understanding is that "Bolt currently works on Windows, Mac OS X, and Linux." (https://godoc.org/github.com/boltdb/bolt) That should cover most rclone users, right? If you are on some other platform, then maybe you just have to live without this feature, at least for now?

**Comment by jediry** [comment]
Yes, thanks for taking this on. Excited to see the improvement!

> garbage collection 
I agree that the user-visibility of this file makes it less of a problem. Still, I could imagine a user with many terabytes of data being unhappy with the "just delete your cache" solution.

If it does become a problem, two possible solutions:
1. Compact-on-shutdown: when terminating rclone after a copy/sync/check, scan the DB for any entries for which you *didn't* look up (or generate) a checksum (for any that you did, you know these are still valid), and check for the existence and mod-time match of the corresponding file. Pro: very automatic, and should be quite cheap when (a) the DB is already well-compacted and (b) most/all the files in it were already examined by the preceding "rclone copy/sync/check" that just finished. Con: potentially significant overhead during shutdown with a large DB, if your "rclone copy/sync/check" operation only examined a small fraction of the whole collection. One way to mitigate this would be to restrict the pruning based on the directory prefix (e.g., if I "rclone" from /mnt/disk1/Media/Movies, then you'd only prune DB entries contained under "/mnt/disk1/Media/Movies"); I believe this minimizes the number of files that need to be examined during compaction.
2. Manual compaction: add "rclone compactdb" (or similar) command, so that the user can compact the cache file following any major directory restructuring. Pro: no overhead in normal usage, can be limited to only running when the user wants it. Con: very manual.

> concurrent use
So long as rclone detects this situation and aborts (rather than continuing and corrupting the DB), I think this is fine. I agree with Monroe-74 that this won't be a huge problem in practice...if I'm routinely processing different directories in parallel, I'll have separate cache DBs for those tasks. The only users likely to be significantly hampered by this are people trying to sync the same local directory into multiple cloud locations...and they can work around this by doing an "rclone check" to bring the DB (mostly) up-to-date, and then making copies of the DB for each "rclone sync" instance they plan to spawn.

> sha1sum, md5sum
Is it fairly cheap to simultaneously compute the SHA1 and MD5 (vs. computing just one)? If it's a significant extra expense to compute both, you might consider making them "optional", and only compute them on-demand. In my case, I'm using OneDrive exclusively, so computing the MD5 sum doesn't matter. I imagine other people's usage is similar. But if it's cheap to do both simultaneously, then probably best to do that, to avoid reading the same file twice to generate two different checksums.

> Use the absolute value of the paths as a key
Depending on what you mean by "absolute value", I probably agree. Specifically, in addition to being a non-relative path, I think the path needs to be canonicalized to remove this-directory (./) and parent-directory (../) sequences, superfluous '/' sequences (e.g., /mnt///path -> /mnt/path), and also should resolve symbolic links.

**Comment by monroe-74** [comment]
Hi jediry, I just want to chime in and say that everything you said makes a lot of sense to me. One minor comment:

<blockquote><p>The only users likely to be significantly hampered by this are people trying to sync the same local directory into multiple cloud locations</p></blockquote>

If I were doing backups to multiple cloud locations, I think I would probably do them on some kind of rolling or alternating schedule, instead of making them simultaneous. I assume a rolling schedule would be a common practice, because then I have more options if I want to rollback to a previous state of my data? So I'm inclined to guess that the exact case you described would be rare. Anyway, you suggested a good way for the user to handle that situation.

**Comment by jediry** [comment]
> If I were doing backups to multiple cloud locations, I think I would probably do them on some kind of rolling or alternating schedule, instead of making them simultaneous.

Yeah, I would do that too. The difficulty is that it's hard to know when a backup task will *finish*, so unless you're _certain_ that the duration of one backup will always be "short" compared to the gap between launching them, then you have to account for the possibility that the previous backup could still be running when the next scheduled "start" rolls around. Of course, it's possible to code around this, but it's more complicated.

**Comment by monroe-74** [comment]
Good point, I see what you mean. I was assuming no overlap, but you're right that this is not a safe assumption.

The new feature we're discussing might radically improve the speed of certain backups, so that might minimize the overlap issue you described. But it will still happen in some situations, so it's good that this problem is getting some thought ahead of time, and I like your suggestions.

**Comment by ncw** [comment]
@jediry @monroe-74 thanks for thinking about this.

I think the garbage collection would be solved for the time being with some extra docs.  If necessary rclone could grow an extra command to help deal with it.  The compact on shutdown is a reasonable idea, but even scanning a huge directory tree can take some time.  We'll have to see how big the DBs become, but pessimistically 1k per file.  So for 1M files the DB becomes 1GB.  So if there are 100k stale files that is not going to be much of a problem.

As for concurrent access, rclone could wait until the db was available rather than stop with an error.  It could print a message every 60s while it was waiting.  This would mean that backups would proceed in an orderly fashion.  boltdb takes a lot of care to not allow concurrent access so you wont get corruption.

I could solve the concurrent access to the db problem like this if necessary
  * the file you pass in as the cache rclone only reads from - multiple readers are allowed if read only
  * rclone writes to a temporary db
  * when rclone has finished the sync it
    * waits for any other users of the main db to finish
    * opens the main db r/w
    * updates any changes

Using multiple DB's is a good idea.  You could fill it up from one run then copy it and use it in another.

> Is it fairly cheap to simultaneously compute the SHA1 and MD5

The main cost is IO so it is pretty much free computing the SHA1 if you are already computing the MD5.  That is what rclone does at the moment anyway.

> Specifically, in addition to being a non-relative path, I think the path needs to be canonicalized to remove this-directory (./) and parent-directory (../) sequences, superfluous '/' sequences (e.g., /mnt///path -> /mnt/path), and also should resolve symbolic links.

Absolutely (pun intended ;-)

**Comment by ncw** [comment]
> and also should resolve symbolic links.

rclone ignores symbolic links at the moment - that needs a little bit of extra thought...

**Comment by monroe-74** [comment]
<blockquote><p>I think the garbage collection would be solved for the time being with some extra docs.</p></blockquote>

Sounds good to me.

<blockquote><p>As for concurrent access, rclone could wait until the db was available rather than stop with an error.</p></blockquote>

Good point. I suggested you just stop, but waiting instead would be much friendlier. In my situation it wouldn't matter, but I could picture another user being annoyed if you stop instead of waiting.

<blockquote><p>I could solve the concurrent access to the db problem like this if necessary</p></blockquote>

Deferring writes this way is a clever solution. However, speaking for myself, this is not needed, and I imagine this could be a noticeable chunk of extra work for you. So I suppose this could be a later enhancement, if many users adopt the new feature and express a need for this enhancement.

<blockquote><p>rclone ignores symbolic links at the moment</p></blockquote>

Speaking again strictly for myself, you ignoring symbolic links does not cause any trouble for me.

**Comment by jediry** [comment]
Ignoring symlinks doesn't bother me either. My comment was based on the incorrect assumption that rclone traverses links.

**Comment by lickdragon** [comment]
pinging this issue to become a participant and get updates on conversation (is there a better way?)

**Comment by robbat2** [comment]
Please also capture the inode number & size, to use for verification.

For a given filename, if ANY of mtime, size, (device_number, inode) are different, don't use it from the cache.

**Comment by tsumare** [comment]
## Paths

Since checksum database cleanup is currently planned as a separate operation, there is no reason to store actual paths at all.  Rather than deal with all the complexities of path resolution, `s3cmd` solves this problem by storing `filesystem id / file inode number / mtime / size,hash`.  I don't know what precisely they do on Windows, but I presume something similar exists.

This method gets around symlinks, file replacement (rename another related file with the same size and modtime), and most other related concerns.  Filename is simply never considered.  If the mtime or size change, the cached checksums are invalidated.  When you run your purge comand, it can simply flag all inodes as "expired", flag all the ones it finds as "unexpired", and clean up at the end.

## md5 vs sha1

As far as this complexity goes, you could just ignore it. Rather than calculating the natural checksum for all remotes every time (dbhash for dropbox? quickxorhash for onedrive business or sharepoint?), simply cache whatever you are using *now*.  If I use a md5 now, calculate it.  If I use a sha1 later, calculate it and store both (if the mtime/size haven't changed).  Users will get what they use, no individual hash has to be calculated more than once, and you don't need special configuration where users tell you up front what they're going to need.

Any other method is likely to either spend a lot of CPU power calculating 4? more? different hashes on large syncs, leave some backends out, or require special user knowledge and intervention.

It's nice to conserve I/O, but most users won't use all four hashes (or more if other backends come in later), and optional "extra_hashes" configuration can be added easily later.

**Comment by ivandeex** [comment]
Related to https://github.com/rclone/rclone/issues/180 (remote metadata caching)

**Comment by ivandeex** [comment]
This request _might_ be implemented by using rclone VFS cache.

**Comment by ivandeex** [comment]
@ncw
This was once a popular request.
Now it attracts zero traffic.
Should we close this (or reopen a new ticket, or whatever)?

**Comment by cwseys** [comment]
Would still be useful!

**Comment by jediry** [comment]
+1

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by mpx** [comment]
Yes, this would be very useful. I regularly use `rclone check` to validate a B2 repo. I've run some crude experiments that show a checksum cache would improve it from 3 hours to a few minutes.

Ideally there would also be an option to only validate a fraction of the local checksums each run (beyond time/size/..). This could be similar to `restic check --read-data-subset`.

**Comment by ivandeex** [comment]
Looks like this feature is still hot and needs attention.
Removed cache backend and VFS labels and retained LocalFS label as it's apparently going to be implemented as a new kind of cache of hashes for local files only.

**Comment by ivandeex** [comment]
Looks like this feature is still hot and needs attention.
It will be implemented as soon as we have a clear plan and maintainer for the feature.
No need for +1 comments anymore!
Removing `Backend: Cache` and `VFS` labels and retained `LocalFS` label as
it's apparently going to be implemented as a new kind of _cache of hashes for local files_ only.

**Comment by ivandeex** [comment]
### Idea

This feature could be implemented by a new wrapping backend, let's call it `hasher` for now. The backend would take the following parameters:

- `remote` (like `alias`)
- `hashsums` (comma/blank separated list)
- `workdir` (default: ~/.cache/rclone/CleanedRemote)
- `max-time` - maximum time to keep hashes

### Algorithm

Algorithm of the new backends's `Object.Hash` could be like this:

1. if requested hash is supported by lower level, just pass it
2. if unsupported, create object `fingerprint` (it will include remote, size, modtime if supported, other first-found supported hash if any)
3. if the strict match is found in workdir, return stored hash
4. if not found, then: purge workdir entry, download object, find hashes and store in workdir

### Notes

We should also handle:

- in-backend copy/move
- delete
- update workdir entries whenever object is fully downloaded
- periodically prune workdir entries if `max-time` is not `off`; this should help against thrashing workdir if a file was removed externally or via lower level backend

### Ways to implement

This algorirhm  could be:

1. implemented as a `hasher` backend
2. integrated into VFS
3. integrated into the `alias` backend
4. integrated into LocalFS (???)

@ncw Thoughts?

**Comment by borkd** [comment]
What are the chances for being able to feed/populate the hash cache out of band, instead of having rclone recalculate it periodically?

Also, while this could require backend-specific code, storing and persisting hashes in a Merkle tree could drastically cut down on remote api calls when syncing larger directory trees.

**Comment by ivandeex** [comment]
> What are the chances for being able to feed/populate the hash cache out of band, instead of having rclone recalculate it periodically?

Answered above.

> Also, while this could require backend-specific code, storing and persisting hashes in a Merkle tree could drastically cut down on remote api calls when syncing larger directory trees.

None of currently supported hashes (md5, sha1, quickxor, etc) support merkle tree. As soon as we have such, e.g. blake3, we'll use merkle tree to build hash from partial range reads.

**Comment by tb582** [comment]
Any updates on this 'feature' looking forward to using it !

**Comment by ivandeex** [comment]
#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by kenmorel** [comment]
Any updates on this feature ?

**Comment by ivandeex** [comment]
Experimental `hasher` beta is published as latest `iva` release at https://github.com/ivandeex/rclone/releases, preliminary documentation is at #5587
